torch/csrc/autograd/python_mode.cpp

```diff
@@ -0,0 +1,44 @@
+#include <torch/csrc/autograd/python_mode.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/autograd/python_variable.h>
+#include <ATen/PythonModeTLS.h>
+
+namespace torch { namespace autograd {
+
+PythonTorchDispatchTypeObject::PythonTorchDispatchTypeObject(
+    PyObject* torch_dispatch_type_object,
+    c10::impl::PyInterpreter* pyinterpreter) {
+  data_ = torch_dispatch_type_object;
+  pyinterpreter_ = pyinterpreter;
+  Py_INCREF(data_);
+}
+
+PythonTorchDispatchTypeObject::~PythonTorchDispatchTypeObject() {
+  Py_DECREF(data_);
```

Comment:
What is guaranteeing that the GIL is held here?

Reply:
The only way to delete one of these objects is to call `PythonMode::exit`*. The only way to call PythonMode::exit is through Python.

Reply:
Er, I'm not sure if that logic is correct. I updated the code to call pyinterpreter_->decref which does hold the GIL (and doesn't seem to deadlock)

Reply:
Given the current implementation yes. But nothing prevents someone from deleting that object from c++ without holding the gil. 
Maybe we want a debug internal assert here just in case we add this in the future?

Reply:
I updated the code to hold the GIL (via a call to pyinterpreter_->decref)

Reply:
Yes, updated the code to hold the GIL (via a call to pyinterpreter_->decref)